types/react-query/index.d.ts

```diff
@@ -36,107 +53,204 @@ export interface QueryOptions<TResult> {
+    | PaginatedQueryLoadingState<TVariables>
+    | PaginatedQueryErrorState<TVariables>
+    | PaginatedQuerySuccessState<TResult, TVariables>;
+
+export interface PaginatedQueryLoadingState<TVariables> extends QueryBaseState<TVariables> {
+    status: 'loading';
+    resolvedData: undefined;
+    latestData: undefined;
+    error: null;
 }
 
-export function prefetchQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: PrefetchQueryOptions<TResult>
-): Promise<TResult>;
+export interface PaginatedQueryErrorState<TVariables> extends QueryBaseState<TVariables> {
+    status: 'error';
+    resolvedData: undefined;
+    latestData: undefined;
+    error: Error;
+}
+
+export interface PaginatedQuerySuccessState<TResult, TVariables> extends QueryBaseState<TVariables> {
+    status: 'success';
+    resolvedData: TResult;
+    latestData: TResult;
+    error: null;
+}
+
+export type InfiniteQueryState<TResult, TVariables> =
+    | InfiniteQuerySuccessState<TResult, TVariables>
+    | InfiniteQueryErrorState<TVariables>
+    | InfiniteQueryLoadingState<TVariables>;
+
+export interface InfiniteQueryLoadingState<TVariables> extends InfiniteQueryBaseState<TVariables> {
+    status: 'loading';
+    data: [];
+    error: null;
+}
+
+export interface InfiniteQueryErrorState<TVariables> extends InfiniteQueryBaseState<TVariables> {
+    status: 'error';
+    data: [];
+    error: Error;
+}
 
-export interface PrefetchQueryOptions<TResult> extends QueryOptions<TResult> {
-    force?: boolean;
+export interface InfiniteQuerySuccessState<TResult, TVariables> extends InfiniteQueryBaseState<TVariables> {
+    status: 'success';
+    data: TResult[];
+    error: null;
+}
+
+export interface InfiniteQueryBaseState<TVariables> {
+    isFetching: boolean;
+    isFetchingMore: boolean;
+    failureCount: number;
+    refetch: (arg?: {
+        variables?: TVariables;
+        merge?: (...args: any[]) => any;
+        disableThrow?: boolean;
+    }) => Promise<any>;
+    fetchMore: (variables?: TVariables) => Promise<any>;
+    canFetchMore: boolean;
 }
 
 export function useMutation<TResults, TVariables extends object>(
     mutationFn: MutationFunction<TResults, TVariables>,
-    mutationOptions?: MutationOptions
-): [MutateFunction<TResults, TVariables>, MutationResult<TResults>];
+    mutationOptions?: MutationOptions<TResults>,
+): [MutateFunction<TResults, TVariables>, MutationState<TResults>];
 
-export type MutationFunction<TResults, TVariables extends object> = (
-    variables: TVariables,
-) => Promise<TResults>;
+export type MutationFunction<TResults, TVariables extends object> = (variables: TVariables) => Promise<TResults>;
 
-export interface MutationOptions {
-    refetchQueries?: Array<string | [string, object]>;
-    refetchQueriesOnFailure?: boolean;
+export interface MutationOptions<TResult> {
+    onSuccess?: (data: TResult) => Promise<any> | undefined;
```

Comment:
these should be | void

Reply:
Should it be `Promise<void> | void`?

Reply:
> `void` should be used as a return type, but not as a parameter type.

https://github.com/Microsoft/dtslint/blob/master/docs/void-return.md

Reply:
```suggestion
    onSuccess?: (data: TResult) => Promise<any> | undefined;
```

Reply:
I think that would count as a return type. It looks like that means func param not generic param

Reply:
@JaceHensley you're right. I've mistook that with other line of code. will fix